presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java

```diff
@@ -509,6 +504,450 @@ private static int batchSize(Connection connection)
+                for (int i = 0; i < batch.size(); i++) {
+                    UUID uuid = batch.get(i);
+                    Optional<UUID> deltaUuid = oldShardUuidsMap.get(uuid);
+                    shardsAndIndexDeleter.delete(shardUuidToId.get(uuid), deltaUuid);
+                }
+                updatedCount += shardsAndIndexDeleter.execute();
+            }
+        }
+        if (updatedCount != oldShardUuidsMap.size()) {
+            throw transactionConflict();
+        }
+
+        return new ShardStats(rowCount, compressedSize, uncompressedSize);
+    }
+
+    /**
+     * For shards and delta
+     *
+     * Insert into `shards`                         for both shards and delta shards
+     * Insert into `shard_nodes`  (non-bucketed)    for both shards and delta shards
+     * Insert into index table                      only for shards
+     */
+    private static void insertShardsAndIndex(long tableId, List<ColumnInfo> columns, Collection<ShardInfo> shards, Map<String, Integer> nodeIds, Handle handle, boolean isDelta)
+            throws SQLException
+    {
+        if (shards.isEmpty()) {
+            return;
+        }
+        boolean bucketed = shards.iterator().next().getBucketNumber().isPresent();
+
+        Connection connection = handle.getConnection();
+        try (IndexInserter indexInserter = new IndexInserter(connection, tableId, columns)) {
+            for (List<ShardInfo> batch : partition(shards, batchSize(connection))) {
+                List<Long> shardIds = insertShards(connection, tableId, batch, isDelta);
+
+                if (!bucketed) {
+                    insertShardNodes(connection, nodeIds, shardIds, batch);
+                }
+
+                if (!isDelta) {
+                    for (int i = 0; i < batch.size(); i++) {
+                        ShardInfo shard = batch.get(i);
+                        Set<Integer> shardNodes = shard.getNodeIdentifiers().stream()
+                                .map(nodeIds::get)
+                                .collect(toSet());
+                        indexInserter.insert(
+                                shardIds.get(i),
+                                shard.getShardUuid(),
+                                shard.getBucketNumber(),
+                                shardNodes,
+                                shard.getColumnStats());
+                    }
+                    indexInserter.execute();
+                }
+            }
+        }
+    }
+
+    /**
+     * For shards
+     *
+     * Select id from `shards` table
+     * - Purpose: 1. check the count as pre-check to avoid conflict 2. get statistics 3. use id to perform update
+     *
+     * Update `shards` table   check delta   verify delete count
+     * Update index table      check delta   verify delete count
+     */
+    private void updateShardsAndIndex(long tableId, Map<UUID, Pair<Optional<UUID>, UUID>> toUpdateShard, Handle handle)
```

Comment:
Same comments around failing early and return value from `ShardsAndIndexUpdater` as in `deleteShardsAndIndexWithDelta`.